### ISSUE TITLE
varnish plugin not showing (varnish version 4)

### ISSUE THREAD

**Issue by vpapas** [netdata/netdata#6149]
Hello,

After upgrading to netdata v1.15.0-29-ge11a4e4 varnish metrics are not showing.

Here is the netadata error log:

> 2019-05-28 17:59:52: python.d INFO: plugin[checker] : varnish : checking
2019-05-28 17:59:52: python.d INFO: plugin[checker] : varnish : source successfully loaded
2019-05-28 17:59:52: python.d INFO: plugin[checker] : varnish : found config file '/usr/lib/netdata/conf.d/python.d/varnish.conf'
2019-05-28 17:59:52: python.d INFO: plugin[checker] : varnish : created 1 job(s) from the config
2019-05-28 17:59:52: python.d ERROR: varnish[varnish] : No output from 'varnishstat'. Is it running? Not enough privileges?
2019-05-28 17:59:52: python.d INFO: plugin[checker] : varnish[varnish] : check failed, skipping the job
2019-05-28 17:59:52: python.d INFO: plugin[checker] : varnish : all jobs failed, skipping module

The following command is working:

> su -s /bin/bash -c "varnishstat -1" netdata

Also, user netdata is in varnish group.

We are running varnish 4 (varnish-4.0.5-1.el7.x86_64).

**Comment by ilyam8** [comment]
Hi @vpapas 	

https://github.com/netdata/netdata/blob/e11a4e4a5dc5231ffb2f58ff05f32156c052766a/collectors/python.d.plugin/varnish/varnish.chart.py#L198-L202

The output from 'varnishstat -1' is empty for some reason.

Try to run it in debug mod
https://docs.netdata.cloud/collectors/python.d.plugin/#how-to-debug-a-python-module

**Comment by vpapas** [comment]
Hello,

Here is the output of the debug:

> /usr/libexec/netdata/plugins.d/python.d.plugin 1 debug varnish trace
2019-05-28 18:42:33: python.d INFO: plugin[main] : using python v2
2019-05-28 18:42:33: python.d INFO: plugin[main] : starting setup
2019-05-28 18:42:33: python.d INFO: plugin[main] : checking for config in ['/etc/netdata', '/usr/lib/netdata/conf.d']
2019-05-28 18:42:33: python.d INFO: plugin[main] : config found, loading config '/usr/lib/netdata/conf.d/python.d.conf'
2019-05-28 18:42:33: python.d INFO: plugin[main] : config successfully loaded
2019-05-28 18:42:33: python.d INFO: plugin[main] : starting checker process (1 module(s) to check)
2019-05-28 18:42:33: python.d INFO: plugin[checker] : starting...
2019-05-28 18:42:33: python.d INFO: plugin[checker] : varnish : checking
2019-05-28 18:42:33: python.d INFO: plugin[checker] : varnish : source successfully loaded
2019-05-28 18:42:33: python.d INFO: plugin[checker] : varnish : found config file '/usr/lib/netdata/conf.d/python.d/varnish.conf'
2019-05-28 18:42:33: python.d INFO: plugin[checker] : varnish : created 1 job(s) from the config
2019-05-28 18:42:33: python.d ERROR: varnish[varnish] : No output from 'varnishstat'. Is it running? Not enough privileges?
2019-05-28 18:42:33: python.d INFO: plugin[checker] : varnish[varnish] : check failed, skipping the job
2019-05-28 18:42:33: python.d INFO: plugin[checker] : varnish : all jobs failed, skipping module
2019-05-28 18:42:33: python.d INFO: plugin[checker] : terminating...
2019-05-28 18:42:33: python.d INFO: plugin[main] : stopping checker process
2019-05-28 18:42:33: python.d INFO: plugin[main] : no modules to run

**Comment by ilyam8** [comment]
> No output from 'varnishstat'

i see we are not logging anything here

https://github.com/netdata/netdata/blob/e11a4e4a5dc5231ffb2f58ff05f32156c052766a/collectors/python.d.plugin/python_modules/bases/FrameworkServices/ExecutableService.py#L36-L37

change it to 

```python
 except TypeError as error:
     self.error('error on decoding line "{0}" : {1}'.format(line,  error)) 
     continue 
```

and run the plugin in debug mode again

___

so you are saying that 
 - `sudo su -s /bin/bash netdata`
 - `varnishstat -1`

works for you?

Nothing was changed in varnish module/ExecutableService in 1.15

**Comment by vpapas** [comment]
Hello again,

I've changed the code but i got the same result:

> bash-4.2$ /usr/libexec/netdata/plugins.d/python.d.plugin 1 debug varnish trace
2019-05-29 10:27:37: python.d INFO: plugin[main] : using python v2
2019-05-29 10:27:37: python.d INFO: plugin[main] : starting setup
2019-05-29 10:27:37: python.d INFO: plugin[main] : checking for config in ['/etc/netdata', '/usr/lib/netdata/conf.d']
2019-05-29 10:27:37: python.d INFO: plugin[main] : config found, loading config '/usr/lib/netdata/conf.d/python.d.conf'
2019-05-29 10:27:37: python.d INFO: plugin[main] : config successfully loaded
2019-05-29 10:27:37: python.d INFO: plugin[main] : starting checker process (1 module(s) to check)
2019-05-29 10:27:37: python.d INFO: plugin[checker] : starting...
2019-05-29 10:27:37: python.d INFO: plugin[checker] : varnish : checking
2019-05-29 10:27:37: python.d INFO: plugin[checker] : varnish : source successfully loaded
2019-05-29 10:27:37: python.d INFO: plugin[checker] : varnish : found config file '/usr/lib/netdata/conf.d/python.d/varnish.conf'
2019-05-29 10:27:37: python.d INFO: plugin[checker] : varnish : created 1 job(s) from the config
2019-05-29 10:27:37: python.d ERROR: varnish[varnish] : No output from 'varnishstat'. Is it running? Not enough privileges?
2019-05-29 10:27:37: python.d INFO: plugin[checker] : varnish[varnish] : check failed, skipping the job
2019-05-29 10:27:37: python.d INFO: plugin[checker] : varnish : all jobs failed, skipping module
2019-05-29 10:27:37: python.d INFO: plugin[checker] : terminating...
2019-05-29 10:27:37: python.d INFO: plugin[main] : stopping checker process
2019-05-29 10:27:37: python.d INFO: plugin[main] : no modules to run
DISABLE

Yes, varnishstat works for user netdata.

> su -s /bin/bash netdata
bash-4.2$ varnishstat -1
MAIN.uptime           12602548         1.00 Child process uptime
MAIN.sess_conn        20325204         1.61 Sessions accepted
MAIN.sess_drop               0         0.00 Sessions dropped
MAIN.sess_fail               0         0.00 Session accept failures
MAIN.sess_pipe_overflow            0         0.00 Session pipe overflow
MAIN.client_req_400                0         0.00 Client requests received, subject to 400 errors
MAIN.client_req_411                0         0.00 Client requests received, subject to 411 errors
MAIN.client_req_413                0         0.00 Client requests received, subject to 413 errors
MAIN.client_req_417                0         0.00 Client requests received, subject to 417 errors
MAIN.client_req             20639867         1.64 Good client requests received
MAIN.cache_hit              20545664         1.63 Cache hits
MAIN.cache_hitpass                 0         0.00 Cache hits for pass
MAIN.cache_miss                48684         0.00 Cache misses
MAIN.backend_conn             142523         0.01 Backend conn. success
MAIN.backend_unhealthy           243         0.00 Backend conn. not attempted
MAIN.backend_busy                  0         0.00 Backend conn. too many
MAIN.backend_fail                122         0.00 Backend conn. failures
MAIN.backend_reuse            195476         0.02 Backend conn. reuses
MAIN.backend_toolate          141611         0.01 Backend conn. was closed
MAIN.backend_recycle          337093         0.03 Backend conn. recycles
MAIN.backend_retry                 1         0.00 Backend conn. retry
MAIN.fetch_head                    4         0.00 Fetch no body (HEAD)
MAIN.fetch_length              73793         0.01 Fetch with Length
MAIN.fetch_chunked             32937         0.00 Fetch chunked
MAIN.fetch_eof                     0         0.00 Fetch EOF
MAIN.fetch_bad                     0         0.00 Fetch bad T-E
MAIN.fetch_close                   0         0.00 Fetch wanted close
MAIN.fetch_oldhttp                 0         0.00 Fetch pre HTTP/1.1 closed
MAIN.fetch_zero                  112         0.00 Fetch zero len body
MAIN.fetch_1xx                     0         0.00 Fetch no body (1xx)
MAIN.fetch_204                     0         0.00 Fetch no body (204)
MAIN.fetch_304                231152         0.02 Fetch no body (304)
MAIN.fetch_failed                  0         0.00 Fetch failed (all causes)
MAIN.fetch_no_thread               0         0.00 Fetch failed (no thread)
MAIN.pools                         2          .   Number of thread pools
MAIN.threads                     200          .   Total number of threads
MAIN.threads_limited               0         0.00 Threads hit max
MAIN.threads_created             200         0.00 Threads created
MAIN.threads_destroyed             0         0.00 Threads destroyed
MAIN.threads_failed                0         0.00 Thread creation failed
MAIN.thread_queue_len              0          .   Length of session queue
MAIN.busy_sleep                   44         0.00 Number of requests sent to sleep on busy objhdr
MAIN.busy_wakeup                  44         0.00 Number of requests woken after sleep on busy objhdr
MAIN.sess_queued                   0         0.00 Sessions queued for thread
MAIN.sess_dropped                  0         0.00 Sessions dropped for thread
MAIN.n_object                      4          .   object structs made
MAIN.n_vampireobject               0          .   unresurrected objects
MAIN.n_objectcore                 12          .   objectcore structs made
MAIN.n_objecthead                 15          .   objecthead structs made
MAIN.n_waitinglist                11          .   waitinglist structs made
MAIN.n_backend                    12          .   Number of backends
MAIN.n_expired                 49005          .   Number of expired objects
MAIN.n_lru_nuked                   0          .   Number of LRU nuked objects
MAIN.n_lru_moved             5220263          .   Number of LRU moved objects
MAIN.losthdr                       0         0.00 HTTP header overflows
MAIN.s_sess                 20325204         1.61 Total sessions seen
MAIN.s_req                  20639867         1.64 Total requests seen
MAIN.s_pipe                        0         0.00 Total pipe sessions seen
MAIN.s_pass                    45503         0.00 Total pass-ed requests seen
MAIN.s_fetch                   94187         0.01 Total backend fetches initiated
MAIN.s_synth                      16         0.00 Total synthethic responses made
MAIN.s_req_hdrbytes        980672122        77.82 Request header bytes
MAIN.s_req_bodybytes         3402981         0.27 Request body bytes
MAIN.s_resp_hdrbytes      8125012187       644.71 Response header bytes
MAIN.s_resp_bodybytes    17787050935      1411.39 Response body bytes
MAIN.s_pipe_hdrbytes               0         0.00 Pipe request header bytes
MAIN.s_pipe_in                     0         0.00 Piped bytes from client
MAIN.s_pipe_out                    0         0.00 Piped bytes to client
MAIN.sess_closed                  33         0.00 Session Closed
MAIN.sess_pipeline                 0         0.00 Session Pipeline
MAIN.sess_readahead                0         0.00 Session Read Ahead
MAIN.sess_herd                447276         0.04 Session herd
MAIN.shm_records          1084644352        86.07 SHM records
MAIN.shm_writes            170167587        13.50 SHM writes
MAIN.shm_flushes                   0         0.00 SHM flushes due to overflow
MAIN.shm_cont                      5         0.00 SHM MTX contention
MAIN.shm_cycles                  438         0.00 SHM cycles through buffer
MAIN.sms_nreq                      0         0.00 SMS allocator requests
MAIN.sms_nobj                      0          .   SMS outstanding allocations
MAIN.sms_nbytes                    0          .   SMS outstanding bytes
MAIN.sms_balloc                    0          .   SMS bytes allocated
MAIN.sms_bfree                     0          .   SMS bytes freed
MAIN.backend_req              337999         0.03 Backend requests made
MAIN.n_vcl                         1         0.00 Number of loaded VCLs in total
MAIN.n_vcl_avail                   1         0.00 Number of VCLs available
MAIN.n_vcl_discard                 0         0.00 Number of discarded VCLs
MAIN.bans                          1          .   Count of bans
MAIN.bans_completed                1          .   Number of bans marked 'completed'
MAIN.bans_obj                      0          .   Number of bans using obj.*
MAIN.bans_req                      0          .   Number of bans using req.*
MAIN.bans_added                    1         0.00 Bans added
MAIN.bans_deleted                  0         0.00 Bans deleted
MAIN.bans_tested                   0         0.00 Bans tested against objects (lookup)
MAIN.bans_obj_killed               0         0.00 Objects killed by bans (lookup)
MAIN.bans_lurker_tested            0         0.00 Bans tested against obj

**Comment by ilyam8** [comment]
Ok, see there is hardcoded timeout (`-t 1`), so module executed `varnishstat -1 -t 1`. Try the command with `-t 1`.

And try this code in python interpreter, this is basically same as collector does

```python
import subprocess
import pprint


command = "varnishstat -1 -t 1".split()
command = "ip ro".split()

p = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE)

data = list()

for line in p.stdout:
    data.append(line.decode('utf-8'))

pprint.pprint(data)
```

**Comment by vpapas** [comment]
Ok, we caught the error:

> bash-4.2$ varnishstat -1 -t 1
(null)
usage: varnishstat [-1lV] [-f field_list] [-n varnish_name] [-N filename] [-w delay]
    -1                           # Print the statistics to stdout.
    -f field_list                # Field inclusion glob
                                 # If it starts with '^' it is used as an exclusion list
    -l                           # Lists the available fields to use with the -f option
    -n varnish_name              # The varnishd instance to get logs from
    -V                           # Display the version number and exit
    -w delay                     # Wait delay seconds between updates.  Default is 1 second. Can also be be used with -1, -x or -j for repeated output.
    -x                           # Print statistics to stdout as XML.
    -j                           # Print statistics to stdout as JSON.

Do you mean to change the varnish.chart.py?

>  if self.instance_name:
            self.command = [varnishstat, '-1', '-n', self.instance_name, '-t', '1']
        else:
            self.command = [varnishstat, '-1', '-t', '1']

**Comment by ilyam8** [comment]
ok, i see, there is no `-t` option in your case

**Comment by ilyam8** [comment]
Hi @vpapas. I beleive the bug is fixed in #6168, could you test it?

**Comment by vpapas** [comment]
Hi @ilyam8,

I've tested it ant it works.
Thank you for your quick response.

Vasso
